VNext

AngelMunoz opened the pull request:
I've been doing a lot of refactoring in this branch in preparation to add the newer and most complex features to perla.

Please don't take anything in this branch as final until I actually say "this is how it's going to be" or something similar.

I mean it is very likely that I won't discard this branch given how much I have invested in it at this point.

Main things to look out in this PR for the moment (I will try to update as soon as I keep going forward)

- Goodbye `Perla.Lib`    
  Perla Lib was a bad intent from my side to eventually expose each part of perla as a library that could be used by third parties, it turns out my code is a mess and won't work
- Hello `Perla.Logger`

  This is a very simple logger to standardize how perla does it's logging, perhaps in the future someone may want to improve it and switch the underlaying mechanism from printf to an actuall logger
- Hello `Perla.PackageManager`

  This is one of the most exciting bits for me, if you want to start playing with importmaps with your own server applications this library is for you, this will allow you to create importmaps that you will be able to write to your server-side templates and take advantace of client side npm-like dependency management without node!
- Hello `Perla.Plugins`
  
  This is the guilty piece that is driving everything, here I will add common code that will be used for Perla plugins, the API is not finalized but you can already see what I have in mind if you take a look at the project, it is very likely that I will also include the [gist dependency manager for F# interactive](https://github.com/AngelMunoz/Gist.DependencyManager.Gist) I recently wrote so one can distribute plugins via gists rather than figuring out another distribution mechanism like the templating engine I wrote initially
- Hello Flan

  Flan is an interesting choice rather than writing it in F#/C# I chose Visual basic to write this piece it still be a CLI tool that will be dedicated to handle importmaps with the `Perla.PacakgeManager` this will be intended to be used individually and to also not depend on having .NET installed (a standalone cli tool) just like Perla binaries


### Missing stuff

- dev, prod and testing dependencies 
    
  The new `Perla.PacakgeManager` allows us to handle JS dependencies better, meaning that if we wanted we could handle dependencies for each environment although the API may be a little bit wonky due to differences between the JSPM Generator Api and the Skypack CDN/API it can be possible to support eventually

- Mounted Directories Virtual File System

  This is other of the driving motivations to add the zio library, it should greatly simplify mounting directories and working with what essentially be available for the dev/build server

- Test runner compatible with [Web Test Runner](https://modern-web.dev/docs/test-runner/overview/)

  Once the virtual file system support is there it will be easier to mount and run JS client side tests in perla which is a must to have for some users



There may be more things to track but for the moment these are basically the goals in mind for this PR hopefully I can get this done soon enough people can start testing around

AngelMunoz committed:
- Move most files to Perla
- Remove Perla.Lib project

AngelMunoz committed:
Add Perla.Logger package

AngelMunoz committed:
- Add Plugins Package

AngelMunoz committed:
- Add PackageManager project
- Add F# Signature Files

AngelMunoz committed:
- update perla sln

AngelMunoz committed:
explore fsmake as fake replacement

AngelMunoz committed:
Set globaljson to .NET6

AngelMunoz committed:
Ensure net6.0

AngelMunoz committed:
add Flan Cli

AngelMunoz committed:
fix optional params

AngelMunoz committed:
Fix missing utilities, optional params
and extensions attribute

AngelMunoz committed:
remove unused project

AngelMunoz committed:
small tweaks

AngelMunoz committed:
add fsi files for the plugin files

AngelMunoz committed:
fix open

AngelMunoz committed:
remove refactored functions

AngelMunoz committed:
add new import map management bits

AngelMunoz committed:
update constants

AngelMunoz committed:
clean public api

AngelMunoz committed:
fix open statement

AngelMunoz committed:
add missing values

AngelMunoz committed:
add fist bits of virtualfs

AngelMunoz committed:
add signature files

AngelMunoz committed:
accomodate to current refactorings

AngelMunoz committed:
remove calcetate stuff

nojaf commented on src/Perla.Plugins/Library.fsi:
```diff
@@ -0,0 +1,158 @@
+/// like HTML, CSS, JS, any other extension is considered a
+/// "Custom" extension
+[<RequireQualifiedAccess; Struct>]
+type FileExtension =
```
This feels a bit unnecessary.
Any extensions as a string would fit the bill here.
You can always use active patterns in your own code if you want to pattern-match this.
As a plugin writer, I wouldn't see much benefit in this structure.

AngelMunoz replied:
Hmm yes, you are probably right I'll consider taking this one out

nojaf commented on src/Perla.Plugins/Library.fsi:
```diff
@@ -0,0 +1,158 @@
+        /// The path where the file changed event was activated
+        /// This value should not be modified between plugin transformations
+        /// because it could break further processing of the sfile
+        originalPath: string
```
If you cannot update this field, I would suggest making a `FileTransformRequest` and `FileTransformResponse` type approach to it.

AngelMunoz replied:
The idea of these plugins is that they will run at build time and when the dev server starts, a file changes, so I don't necessarily want to introduce something that ties to dev/build time specifically I'd like to have it agnostic

The plugin should be able to run without caring who is calling it at what time, since I will have control when these are called I might as well ensure that the originalPath stays as is even if users modify this value

But! that being said I'd like to know what do you have in mind for that, could you elaborate a little bit more on this please?

nojaf replied:
Just something like:
```fsharp
type FileTransformRequest =
   { OriginalPath: string
     Content: string
     Extension: string }

type FileTransformResponse =
  { Content: string
    Extension: string }
```

Given that you mentioned `originalPath` should not be updated.

AngelMunoz commented:
> have you considered using FSharp.Adaptive instead of Observables?

Not really, since the file system events I monitor are observables that's why I've been sticking with them, I'm not sure if theres a way I could incorporate adaptive stuff into it

I'll ask Jimmy about it

AngelMunoz committed:
remove extensions file
and move path extensions to file system
take out even more out of this file

AngelMunoz committed:
move env to its own file

AngelMunoz committed:
move json to its own file
Allow write parts of the perla config rather than the whole file

AngelMunoz committed:
Cleanup File system
- only use disk apis in this module
- use static class for a better public api (even if its internal)

AngelMunoz committed:
accomodate to refactors

AngelMunoz committed:
move constants to its own file

AngelMunoz committed:
add dependencies to perla config

AngelMunoz committed:
add and remove files

AngelMunoz committed:
fix up refactor changes

AngelMunoz committed:
go back to 0 until it's time

AngelMunoz committed:
add helper member

AngelMunoz committed:
remove unnecessary optionals from core types
levearage options for that

AngelMunoz committed:
Refactored massively the command execution
remove Result<'T, exn> from everywhere
Cleanup Optional arguments

AngelMunoz committed:
update constants

AngelMunoz committed:
move esbuild file ops to file system

AngelMunoz committed:
move functions from commands to the library file
not sure where to put them still

AngelMunoz committed:
removed unused json method

AngelMunoz committed:
Attempt to centralize configuration access
And to allow merging  options from env vars -> json file -> cli
God I hate Json

AngelMunoz commented:
Hello watcher folks, I hope you have been having a great week so far, here's the next batch of news:

This last set of changes includes most of the major refactoring to eliminate some of the major pains when maintaining this project, the optional values in the core types, to alleviate that I wrote a Configuration module that hopefully should centralize access to the configuration and also a single point for also updating the perla config file

New aditions to the configuration system also allows us to have preference order

Default settings -> Environment Variables -> Config file -> CLI overrides

This also means that perla should be able to run without a config file as long as one has an index file and an src directory :)

While the environment vars change is not enabled yet we can easily support that once we settle which (and if we want) fields to allow being set from env vars, that will likely be used for feature flags at some point.

The first major breaking chances for the config file are in, I've sourced some of the options to be top level ones like runConfiguration (production/development), moved the env file options also to be top level and introduced a few new ones like dependencies and devDependencies!

This change will allow you as a consumer to load dependencies that work with browser extensions like react dev tools or similar yet they won't be there when your production code is built

The FileSystem module is also now the central point for File/directory operations, While this was the goal behind the Fs module in previous iterations, the file system operations were still scattered all over the codebase, this should be in better shape now

The Scaffolding module as well is now self contained to all of the templating operations these may have broken while I did the refactoring, but with the introduction of UMX to the code base I will be able to ensure that I'm not confusing magic strings all over the place the functionality should not change at all so I will have to do the due testing in time.

We moved on from Argu to https://github.com/JordanMarr/FSharp.SystemCommandLine, which offers a different way of parsing CLI arguments, the main reasoning behind it is to also leverage tab completitons, a more customizable CLI for users, and the integration to IHost and some DI utils which may be quite useful within the aspnet server while the commands have been ported, We still need to ensure these commands work better than before

There's still pending refactoring in the following modules

- Fable
- Esbuild
- Server
- Build

#### Fable

I don't think the Fable module will have significant changes other than allow passing the source maps flag to the fable command and to accomodate to the previous optional settings that should be set this time.

#### Esbuild

Similar story to the Fable module, it is already pretty self contained and works nice, what could happen here is that we may remove existing compilation functions in favor to create an in-source Esbuild plugin for the Blessed types (JS, TS, JSX, TSX, CSS) but that will be after the virtual file system and plugin stuff is introduced after the refactoring efforts are done.

#### Server and Build

These two will be the last and are the most critical modules to get right as the premise of Perla lays in them they will be reworked after the new virtual file system and plugins work is introduced in the code base

The plan is to be able to leverage the mounted modules with the zio library to have a better path integration with both source code paths and server paths

The idea is that once the virtual file system is correctly mapped, we'll run each existing file in the mounted directories through the plugin list this in turn should leave us with the end products HTML/JS/CSS which we will process with esbuild in the end to generate a single bundle of JS/CSS and to emit the index file with the correct import map

At dev (or also reffered as perla serve) time we'll do the exact same and the files should be ready to be served by the aspnet app, each time a file changes we will be able to fire of an event either with observables or with adaptive values (I did speak with Jimmy about it and it might be useful for us as well, thanks nojaf for the pointer) and send that individual file through the pipeline and place it in the virtual file system. This should not introduce any noticeable speed changes which should be as fast as today

And I think that's what I got for today.

An extra note in the breaking chances is mostly in the configuration file not in the apps themselves moving/removing/adding some properties should be enough to fix those in your projects.

AngelMunoz committed:
bump dependencies

AngelMunoz committed:
split build and esbuild options in two

AngelMunoz committed:
update default esbuild version

AngelMunoz committed:
update commands to refactors

AngelMunoz committed:
refactor esbuid/fable to new type changes
also move esbuild above to allow authoring a plugin with the module

AngelMunoz committed:
add try read tsconfig
add Fake globbing

AngelMunoz committed:
add first bits of plugin infra

AngelMunoz committed:
update to refactors

AngelMunoz committed:
add first bits of Virtual System

AngelMunoz committed:
update to current refactors and add fsi files

AngelMunoz committed:
Still broken, will fix soon

AngelMunoz committed:
logger changes nothing fancy

AngelMunoz committed:
fix typo

AngelMunoz committed:
Add Copy MountedDirs to Disk

add a public surface mount with the perla configuration

AngelMunoz committed:
change prop to version

AngelMunoz committed:
Enable public Esbuild class

fix compilation issues with esbuild file
switch the esbuild pipelines to the esbuild plugin
cleanup optional types

AngelMunoz committed:
add temp dir function

AngelMunoz committed:
Cleanup optional types

Cleaned up optional types and  revamped the build process
- Mount the virtual system (plugins will run in this step)
- Copy the File System to disk to a temporary directory
- Run esbuild on the temp dir and output to outDir
- inject css and importmap
- emit env file if enabled
- copy files with fable globs

AngelMunoz committed:
Took out fable compilation and outdir cleanup

AngelMunoz committed:
update file positioning

AngelMunoz committed:
run operations in the same call

AngelMunoz committed:
add format command

AngelMunoz committed:
Run Fantomas

AngelMunoz committed:
include fsi files

AngelMunoz committed:
rename file add back dependency

AngelMunoz committed:
slight re-work on the fable file

AngelMunoz committed:
slight changes to the logger utils

AngelMunoz committed:
rework server stuff
- remove all of the compilation stuff
- focus on serving files
- externalize calls make a single entry point with the
  Server type

AngelMunoz committed:
fix compile issues due refactor

AngelMunoz committed:
move docs to root
move experiments to root
move app to root

AngelMunoz committed:
add fsi files

AngelMunoz committed:
switch to paket
add tests project
add update tools.fsx

AngelMunoz committed:
add thoth.json

AngelMunoz committed:
fix config reads from file

AngelMunoz committed:
fix compile issue

AngelMunoz committed:
remove conflicting command

AngelMunoz committed:
fix start command

AngelMunoz committed:
add fist test :fire:

AngelMunoz committed:
half fix server to at least make it run

AngelMunoz committed:
format

AngelMunoz committed:
format

AngelMunoz committed:
add internal api

AngelMunoz committed:
format and add env tests

AngelMunoz committed:
fix missing value

AngelMunoz committed:
add a couple of json tests

AngelMunoz committed:
fantomas and rename tests to env

AngelMunoz committed:
change module

AngelMunoz committed:
add esbuild json tests

AngelMunoz committed:
update tests

AngelMunoz committed:
update docs to accomodate to vnext manual testing

AngelMunoz committed:
fix up missing plugins

AngelMunoz committed:
fantomas

AngelMunoz committed:
fix css compile in esbuild plugin

AngelMunoz committed:
- fix missing plugins
- allow to check against plugin predicates

AngelMunoz committed:
add cancellation token support

AngelMunoz committed:
fix setting perla cwd

AngelMunoz committed:
re-enable old behavior for css/json
return mimetypes for known files

AngelMunoz committed:
fix a couple of misplaced files during build

AngelMunoz committed:
ensure at least setting cwd is working

AngelMunoz committed:
add esbuild tests

AngelMunoz committed:
update debug configs for vscode

AngelMunoz committed:
update perla config and importmap in the docs dir

AngelMunoz committed:
update debug configs for rider/vs2022

AngelMunoz committed:
Allowed regenerate to be more flexible with res

AngelMunoz committed:
let the user know what's going on

AngelMunoz committed:
add provider from cli args

AngelMunoz committed:
ad cancellation token support
for esbuild downloads

AngelMunoz committed:
update dependency fields

AngelMunoz committed:
get static dependencies

AngelMunoz committed:
update commands in program

AngelMunoz committed:
fix css/json module loading

AngelMunoz committed:
fix up build command
cleanup build module
add log indicators for builds

AngelMunoz committed:
fix failing test

AngelMunoz committed:
remove --no-build flag

AngelMunoz committed:
update sdk

AngelMunoz committed:
.

AngelMunoz committed:
rename a couple of namespaces/files

AngelMunoz committed:
add build tests

AngelMunoz committed:
add depenencies tests

AngelMunoz committed:
add tests for types

AngelMunoz committed:
switch from struct to normal record and options

AngelMunoz committed:
update run config

AngelMunoz committed:
update program and rename commands

AngelMunoz committed:
fix writing to disk

AngelMunoz committed:
ensure add/remove only update
wanted dependencies in perla.jsonc

AngelMunoz committed:
write perlaconfig to disk

AngelMunoz committed:
fix add commands
fix remove commands
ensure restore works fine

AngelMunoz committed:
add tests for new method

AngelMunoz committed:
fix missing fable configuraton
when reading froma  file

AngelMunoz committed:
fix typo

AngelMunoz committed:
add serilog to aspnet server

AngelMunoz committed:
update vscode debug settings

AngelMunoz committed:
update sample to vnext

AngelMunoz committed:
make JS files go through without esbuild

AngelMunoz committed:
add serilog for cleaner output

AngelMunoz committed:
fix workflows for build and serves

AngelMunoz commented:
Hello there Folks!

The next batch of changes are in!

This week the focus was on cleaning up internal APIs, ensuring things build/serve as they should, the internal esbuild plugin works and leave things in shape for what comes after v1.0.0

Here is the updates:

- #42 has been finally addressed...
    
  while the test surface area is big and the most complex parts may not be that testable (bad design skills sorry) there are some tests there that check import map generation, some json parsing (to read from confg and other parts of the system, and most important leaves the testing stuff open for others to contribute if desired

- #59 has also been addressed
  
  perla.jsonc is basically not needed in disk for perla to work, while we do need it to track dependencies the amount of required stuff inside it has been drastically required

- #77 We added fake globbing to solve this issue
- #79 has landed as well

  I plan to add a [test runner](https://modern-web.dev/docs/test-runner/overview/) for client-side tests and playwright, so this was needed to be done eventually

- #80 has landed as well
- #88 should be solved as well

The first previews will start going out during the next week or two so stay tuned

This includes support for plugins!

once the packages are out, I'll let you know 

that's it for the meantime, I'll see you in the next one!

AngelMunoz committed:
remove the phrase from the docs

AngelMunoz committed:
Fix scaffolding commands

AngelMunoz committed:
fix tests

AngelMunoz committed:
add new options to the readme

AngelMunoz committed:
remove the Nacre experiment

AngelMunoz committed:
Add Playwright to Perla deps

AngelMunoz committed:
formatting

AngelMunoz committed:
remove unnecessary qualification

AngelMunoz committed:
add testing helpers, and mocha runner

AngelMunoz committed:
add testing helpers and mocha runner JS files

AngelMunoz committed:
add testing types and decoders

AngelMunoz committed:
add testing module

AngelMunoz committed:
refactor services and middleware

add WebApplication pipeline for Dev and Test apps

AngelMunoz committed:
add test command preliminary bits

AngelMunoz committed:
update GH CI pipeline

AngelMunoz committed:
update pipeline versions

AngelMunoz committed:
add paket restore

AngelMunoz committed:
make ci work this time?

AngelMunoz committed:
- Make project testable
- Add Sample Tests

AngelMunoz committed:
Fix adding -d not being respected by the config manager

AngelMunoz committed:
ensure mocha is in window

AngelMunoz committed:
attempt to get unstuck

AngelMunoz committed:
Format

AngelMunoz committed:
download esbuild separately for CI

AngelMunoz committed:
Ensure Fable runs first in both watch/non-watch mode when running
- build
- serve
- test

AngelMunoz committed:
simplify paket

AngelMunoz committed:
update tools and run fantomas

AngelMunoz committed:
fix a couple o bugsrelated to the test files

AngelMunoz committed:
prepare first v1 pre-release

AngelMunoz commented:
Okay Folks! It is time for v1 pre-releases otherwise I will never get feedback and I will never keep it going because of not enough motivation to finish the testing stuff and we'll go into a vicious cycle

v1 pre-releases will hit NuGet soon!
